dynamicio/mixins/with_kafka.py

```diff
@@ -124,39 +297,39 @@ def _get_producer(self, server: str, **options: MutableMapping[str, Any]) -> Kaf
             A Kafka producer instance.
         """
         self.__kafka_config = {
-            **{
-                "bootstrap_servers": server,
-                "compression_type": "snappy",
-                "key_serializer": self._default_key_serializer,
-                "value_serializer": self._default_value_serializer,
-            },
+            "bootstrap.servers": server,
+            "compression.type": "snappy",
             **options,
         }
-        return KafkaProducer(**self.__kafka_config)
+        return Producer(**self.__kafka_config)
 
     def _send_messages(self, df: pd.DataFrame, topic: str) -> None:
-        logger.info(f"Sending {len(df)} messages to Kafka topic:{topic}.")
+        logger.info(f"Sending {len(df)} messages to Kafka topic: {topic}.")
 
         messages = df.reset_index(drop=True).to_dict("records")
+
         for idx, message in zip(df.index.values, messages):
-            self.__producer.send(topic, key=self.__key_generator(idx, message), value=self.__document_transformer(message))  # type: ignore
+            key = self.__key_generator(idx, message)
+            transformed_message = self.__document_transformer(message)
+            serialized_key = self.__key_serializer(key)
+            serialized_value = self.__value_serializer(transformed_message)
+
+            self.__producer.produce(topic=topic, key=serialized_key, value=serialized_value, on_delivery=self._on_delivery)
 
-        self.__producer.flush()  # type: ignore
+        self.__producer.flush()
 
     @staticmethod
-    def _default_key_serializer(key: Optional[str]) -> Optional[bytes]:
-        if key:
-            return key.encode("utf-8")
+    def _on_delivery(err, msg):
+        """Callback for message delivery."""
+        if err is not None:
+            raise Exception(f"Message delivery failed: {err}, for message: {msg}")
```

Comment:
Single message failing to be sent will trow an exception as demonstrated below: 
```shell
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Traceback (most recent call last):\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/.pyenv/versions/3.10.4/lib/python3.10/multiprocessing/pool.py", line 125, in worker\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    result = (True, func(*args, **kwds))\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 130, in send_voyage_documents\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    raise Exception(error_msg) from generic_exception\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Exception: Exception: Message delivery failed: KafkaError{code=NOT_ENOUGH_REPLICAS,val=19,str="Broker: Not enough in-sync replicas"}, for message: <cimpl.Message object at 0x7fc2ed58ae40>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Traceback (most recent call last):\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 82, in send_voyage_documents\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    kafka_io_write(wrapped_documents, mode)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 171, in kafka_io_write\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    _connect_with_retry(lambda: kafka_io.write(pd.DataFrame.from_records(documents_as_dicts).set_index("key")), _KAFKA_CONNECTION_ATTEMPTS, _KAFKA_CONNECTION_BACKOFF_SECS)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 136, in _connect_with_retry\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    return write_op()\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 171, in <lambda>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    _connect_with_retry(lambda: kafka_io.write(pd.DataFrame.from_records(documents_as_dicts).set_index("key")), _KAFKA_CONNECTION_ATTEMPTS, _KAFKA_CONNECTION_BACKOFF_SECS)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/core.py", line 185, in write\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    getattr(self, f"_write_to_{source_name}")(self._apply_schema(df))\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 259, in _write_to_kafka\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    self._send_messages(df=df, topic=self.sources_config.kafka.kafka_topic)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 319, in _send_messages\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    self.__producer.flush()\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 325, in _on_delivery\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    raise Exception(f"Message delivery failed: {err}, for message: {msg}")\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Exception: Message delivery failed: KafkaError{code=NOT_ENOUGH_REPLICAS,val=19,str="Broker: Not enough in-sync replicas"}, for message: <cimpl.Message object at 0x7fc2ed58ae40>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'"""\n'
```

There is room to relax this condition with a parameter that controls how sensitive the application should be. I.e. throw exceptions after 10-20-100 failed deliveries, else just log them... 

I think the library should be super-sensitive to message sending though, hence why I implemented it like this.